File: python/dmp_discrete/FuncApproximatorDiscrete.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
import copy
sys.path.append(os.path.join(os.path.dirname(__file__), "../dmp_param/"))
sys.path.append(os.path.join(os.path.dirname(__file__), "../dmp_base/"))
from FunctionApproximator import *
from CanonicalSystemDiscrete import *


class FuncApproximatorDiscrete(FunctionApproximator, object):
  "Class for function approximators of discrete DMPs."

  def __init__(self,
               dmp_num_dimensions_init,
               num_basis_functions,
               canonical_system_discrete,
               name=""):
    super(FuncApproximatorDiscrete,
          self).__init__(dmp_num_dimensions_init, num_basis_functions,
                         canonical_system_discrete, name)
    self.centers = np.zeros((self.model_size, 1))
    self.bandwidths = np.zeros((self.model_size, 1))
    self.psi = np.zeros((self.model_size, 1))
    self.initBasisFunctions()

  def initBasisFunctions(self):
    canonical_order = self.canonical_sys.order
    assert (
        (canonical_order == 1) or (canonical_order == 2)
    ), "Discrete canonical system order=" + str(canonical_order) + (
        " is not supported! The only supported discrete canonical system order"
        " is either order==1 or order==2.")
    alpha_canonical = self.canonical_sys.alpha

    t = np.array(np.linspace(0.0, 1.0, self.model_size)).reshape(
        (self.model_size, 1)) * self.canonical_sys.tau_sys.tau_reference
    if (canonical_order == 2):
      self.centers = (1.0 + (
          (alpha_canonical / 2.0) * t)) * np.exp(-(alpha_canonical / 2.0) * t)
    elif (canonical_order == 1):
      self.centers = np.exp(-alpha_canonical * t)
    self.bandwidths = np.square(np.diff(self.centers, axis=0) * 0.55)
    self.bandwidths = 1.0 / np.vstack((self.bandwidths, self.bandwidths[-1, 0]))

    return None

  def isValid(self):
    assert (super(FuncApproximatorDiscrete, self).isValid())
    assert (self.canonical_sys.isValid())
    assert (self.centers.shape[0] == self.model_size)
    assert (self.bandwidths.shape[0] == self.model_size)
    assert (self.psi.shape[0] == self.model_size)
    assert (np.array_equal(self.centers, np.zeros(
        (self.model_size, 1))) == False)
    assert (np.array_equal(self.bandwidths, np.zeros(
        (self.model_size, 1))) == False)
    return True

  def getBasisFunctionTensor(self, canonical_X):
    assert (self.isValid())

    if (type(canonical_X) != np.ndarray):
      cX = canonical_X * np.ones((1, 1))
    else:
      assert (canonical_X.shape[0] == 1), "canonical_X must be a row vector!"
      cX = canonical_X
    basis_function_tensor = np.exp(
        -0.5 * np.square(
            np.matmul(np.ones(self.centers.shape), cX) -
            np.matmul(self.centers, np.ones(cX.shape))) *
        np.matmul(self.bandwidths, np.ones(cX.shape)))
    return basis_function_tensor

  def getNormalizedBasisFunctionVectorMultipliedPhaseMultiplier(self):
    assert (self.isValid())

    self.psi = self.getBasisFunctionTensor(
        self.canonical_sys.getCanonicalPosition())
    assert (self.psi.shape == (
        self.model_size,
        1)), "self.psi must be a column vector of size self.model_size!"
    assert (np.isnan(self.psi).any() == False), "self.psi contains NaN!"
    sum_psi = np.sum(self.psi + 1.e-10)
    normalized_basis_func_vector_mult_phase_multiplier = self.psi * self.canonical_sys.getCanonicalMultiplier(
    ) * 1.0 / sum_psi
    assert (np.isnan(
        normalized_basis_func_vector_mult_phase_multiplier).any() == False
           ), "normalized_basis_func_vector_mult_phase_multiplier contains NaN!"

    return normalized_basis_func_vector_mult_phase_multiplier

  def getForcingTerm(self):
    assert (self.isValid())

    normalized_basis_func_vector_multiplied_phase_multiplier = self.getNormalizedBasisFunctionVectorMultipliedPhaseMultiplier(
    )
    forcing_term = np.matmul(
        self.weights, normalized_basis_func_vector_multiplied_phase_multiplier)
    assert (np.isnan(forcing_term).any() == False), "forcing_term contains NaN!"
    basis_function_vector = copy.copy(self.psi)

    return forcing_term, basis_function_vector

  def getForcingTermTraj(self, canonical_position_trajectory,
                         canonical_velocity_trajectory):
    assert (self.isValid())

    traj_length = canonical_position_trajectory.shape[1]
    X = canonical_position_trajectory
    V = canonical_velocity_trajectory
    PSI = self.getBasisFunctionTensor(X)
    assert (PSI.shape[0] == self.model_size)
    assert (np.isnan(PSI).any() == False), "PSI contains NaN!"
    if (self.canonical_sys.order == 1):
      forcing_term_trajectory = np.matmul(self.weights, PSI) * np.matmul(
          np.ones((self.dmp_num_dimensions, 1)),
          (X * 1.0 / np.sum(PSI, axis=0).reshape((1, traj_length))))
      #forcing_term_trajectory = np.matmul(self.weights, PSI) * np.matmul(np.ones((self.dmp_num_dimensions, 1)), (X * 1.0 / np.sum(PSI + 1.e-10, axis=0).reshape((1,traj_length))))
    elif (self.canonical_sys.order == 2):
      forcing_term_trajectory = np.matmul(self.weights, PSI) * np.matmul(
          np.ones((self.dmp_num_dimensions, 1)),
          (V * 1.0 / np.sum(PSI, axis=0).reshape((1, traj_length))))
      #forcing_term_trajectory = np.matmul(self.weights, PSI) * np.matmul(np.ones((self.dmp_num_dimensions, 1)), (V * 1.0 / np.sum(PSI + 1.e-10, axis=0).reshape((1,traj_length))))
    assert (np.isnan(forcing_term_trajectory).any() == False
           ), "forcing_term_trajectory contains NaN!"
    basis_function_trajectory = PSI

    return forcing_term_trajectory, basis_function_trajectory

File: src/dmp_discrete/FuncApproximatorDiscrete.cpp
#include "dmp/dmp_discrete/FuncApproximatorDiscrete.h"

namespace dmp {

FuncApproximatorDiscrete::FuncApproximatorDiscrete()
    : FunctionApproximator(),
      centers(new VectorM(model_size)),
      bandwidths(new VectorM(model_size)),
      psi(new VectorM(model_size)),
      sum_psi(0.0) {}

FuncApproximatorDiscrete::FuncApproximatorDiscrete(
    uint dmp_num_dimensions_init, uint num_basis_functions,
    CanonicalSystemDiscrete* canonical_system_discrete,
    RealTimeAssertor* real_time_assertor)
    : FunctionApproximator(dmp_num_dimensions_init, num_basis_functions,
                           canonical_system_discrete, real_time_assertor),
      centers(new VectorM(model_size)),
      bandwidths(new VectorM(model_size)),
      psi(new VectorM(model_size)),
      sum_psi(0.0),
      canonical_sys_discrete(canonical_system_discrete) {
  initBasisFunctions();
}

void FuncApproximatorDiscrete::initBasisFunctions() {
  uint canonical_order = getCanonicalSystemDiscreteOrder();
  double alpha_canonical = canonical_sys_discrete->getAlpha();
  double tau_reference =
      canonical_sys->getTauSystemPointer()->getTauReference();

  // centers are spanned evenly within <tau_reference> (normally 0.5 seconds)
  // period of the evolution of the canonical state position:
  double dt = ((1.0 - 0.0) / (model_size - 1)) * tau_reference;

  // distribute Gaussian centers within 0.5 seconds period of
  // the decaying-exponential evolution of the canonical state position:
  for (uint m = 0; m < model_size; m++) {
    double t = m * dt;
    if (canonical_order == 2) {
      // analytical solution to differential equation:
      // tau^2 * ydd(t) = alpha * ((beta * (0.0 - y(t))) - (tau * yd(t)))
      // with initial conditions y(0) = 1.0 and yd(0) = 0.0
      // beta = alpha / 4.0 (for critical damping response on the 2nd order
      // system) tau = 1.0 and centers[m] = y(m)
      (*centers)[m] = (1.0 + ((alpha_canonical / 2.0) * t)) *
                      exp(-(alpha_canonical / 2.0) * t);
    } else {  // 1st order canonical system
      // analytical solution to differential equation:
      // tau * yd(t) = -alpha * y(t)
      // with initial conditions y(0) = 1.0
      // tau = 1.0 and centers[m] = y(m)
      (*centers)[m] = exp(-alpha_canonical * t);
    }
  }

  // Define bandwidths around computed centers
  for (uint i = 0; i < (model_size - 1); i++) {
    // (*bandwidths)[i]    = 1.0 / (((*centers)[i+1] - (*centers)[i]) *
    // ((*centers)[i+1] - (*centers)[i]));                // original chebotar's
    //  implementation
    (*bandwidths)[i] =
        1.0 / (0.55 * ((*centers)[i + 1] - (*centers)[i]) * 0.55 *
               ((*centers)[i + 1] -
                (*centers)[i]));  // gsutanto: following Stefan's implementation
                                  // in dcp.m (better overlap between
                                  // neighboring basis functions)
  }
  (*bandwidths)[model_size - 1] = (*bandwidths)[model_size - 2];
}

bool FuncApproximatorDiscrete::isValid() {
  if (rt_assert(FunctionApproximator::isValid()) == false) {
    return false;
  }
  if (rt_assert(canonical_sys_discrete->isValid()) == false) {
    return false;
  }
  if (rt_assert((rt_assert(centers != nullptr)) &&
                (rt_assert(bandwidths != nullptr)) &&
                (rt_assert(psi != nullptr))) == false) {
    return false;
  }
  if (rt_assert((rt_assert(centers->rows() == model_size)) &&
                (rt_assert(bandwidths->rows() == model_size)) &&
                (rt_assert(psi->rows() == model_size))) == false) {
    return false;
  }
  return true;
}

bool FuncApproximatorDiscrete::getForcingTerm(
    VectorN& result_f, VectorM* basis_function_vector,
    VectorM* normalized_basis_func_vector_mult_phase_multiplier) {
  // pre-condition checking
  if (rt_assert(FuncApproximatorDiscrete::isValid()) == false) {
    return false;
  }
  // input checking
  if (rt_assert(result_f.rows() == dmp_num_dimensions) == false) {
    return false;
  }

  VectorM normalized_basis_func_vector_multiplied_phase_multiplier =
      ZeroVectorM(model_size);
  if (rt_assert(getNormalizedBasisFunctionVectorMultipliedPhaseMultiplier(
          normalized_basis_func_vector_multiplied_phase_multiplier)) == false) {
    return false;
  }

  if (basis_function_vector != nullptr) {
    if (rt_assert(basis_function_vector->rows() == model_size) == false) {
      return false;
    }
    *basis_function_vector = *psi;
  }

  if (normalized_basis_func_vector_mult_phase_multiplier != nullptr) {
    (*normalized_basis_func_vector_mult_phase_multiplier) =
        normalized_basis_func_vector_multiplied_phase_multiplier;
  }

  VectorN f = ZeroVectorN(dmp_num_dimensions);
  f = ((*weights) * normalized_basis_func_vector_multiplied_phase_multiplier);

  if (rt_assert(containsNaN(f) == false) == false) {
    return false;
  }

  result_f = f;

  return true;
}

bool FuncApproximatorDiscrete::getFuncApproxLearningComponents(
    VectorM& current_psi, double& current_sum_psi, double& current_xi) {
  // pre-condition checking
  if (rt_assert(FuncApproximatorDiscrete::isValid()) == false) {
    return false;
  }
  // input checking
  if (rt_assert(current_psi.rows() == model_size) == false) {
    return false;
  }

  // Update basis functions vector (psi) and its sum (sum_psi)
  if (rt_assert(getBasisFunctionVector(canonical_sys->getCanonicalPosition(),
                                       (*psi))) == false) {
    return false;
  }
  if (rt_assert(sum_psi != 0.0) == false) {
    return false;
  }

  current_psi = (*psi);
  current_sum_psi = sum_psi;
  current_xi = (canonical_sys->getCanonicalMultiplier());

  return true;
}

bool FuncApproximatorDiscrete::
    getNormalizedBasisFunctionVectorMultipliedPhaseMultiplier(
        VectorM& normalized_basis_func_vector_mult_phase_multiplier) {
  // pre-condition checking
  if (rt_assert(FuncApproximatorDiscrete::isValid()) == false) {
    return false;
  }
  // input checking
  if (rt_assert(normalized_basis_func_vector_mult_phase_multiplier.rows() ==
                model_size) == false) {
    return false;
  }

  // Update basis functions vector (psi) and its sum (sum_psi)
  if (rt_assert(getBasisFunctionVector(canonical_sys->getCanonicalPosition(),
                                       (*psi))) == false) {
    return false;
  }
  if (rt_assert(sum_psi != 0.0) == false) {
    return false;
  }

  normalized_basis_func_vector_mult_phase_multiplier =
      (*psi) * ((canonical_sys->getCanonicalMultiplier()) / sum_psi);

  return true;
}

bool FuncApproximatorDiscrete::getBasisFunctionVector(double x,
                                                      VectorM& result_vector) {
  // pre-condition checking
  if (rt_assert(FuncApproximatorDiscrete::isValid()) == false) {
    return false;
  }
  // input checking
  if (rt_assert(result_vector.rows() == model_size) == false) {
    return false;
  }

  // Compute evaluation of each Gaussian
  for (uint i = 0; i < model_size; ++i) {
    (*psi)[i] = getBasisFunctionValue(x, (*centers)[i], (*bandwidths)[i]);
  }
  //    sum_psi         = (psi->colwise().sum())(0,0);
  sum_psi = (psi->colwise().sum())(0, 0) + (model_size * 1.e-10);

  result_vector = (*psi);

  return true;
}

double FuncApproximatorDiscrete::getBasisFunctionValue(double x, double center,
                                                       double bandwidth) {
  double temp = (x - center);
  return (exp(-0.5 * temp * temp * bandwidth));
}

bool FuncApproximatorDiscrete::getWeights(MatrixNxM& weights_buffer) {
  // pre-condition checking
  if (rt_assert(FuncApproximatorDiscrete::isValid()) == false) {
    return false;
  }

  weights_buffer = *weights;
  return true;
}

bool FuncApproximatorDiscrete::setWeights(const MatrixNxM& new_weights) {
  // pre-condition checking
  if (rt_assert(FuncApproximatorDiscrete::isValid()) == false) {
    return false;
  }
  // input checking
  if (rt_assert((rt_assert(new_weights.rows() == dmp_num_dimensions)) &&
                (rt_assert(new_weights.cols() == model_size))) == false) {
    return false;
  }

  *weights = new_weights;
  return true;
}

uint FuncApproximatorDiscrete::getCanonicalSystemDiscreteOrder() {
  return canonical_sys_discrete->getOrder();
}

double FuncApproximatorDiscrete::getCanonicalSystemDiscretePosition() {
  return canonical_sys_discrete->getX();
}

double FuncApproximatorDiscrete::getCanonicalSystemDiscreteVelocity() {
  return canonical_sys_discrete->getXd();
}

double FuncApproximatorDiscrete::getCanonicalSystemDiscreteAcceleration() {
  return canonical_sys_discrete->getXdd();
}

FuncApproximatorDiscrete::~FuncApproximatorDiscrete() {}

}  // namespace dmp
